modules/apstreamBidAdapter.js

```diff
@@ -0,0 +1,484 @@
+    return getConsentStringFromPrebid(bidderRequest.gdprConsent);
+  }
+
+  return 'disabled';
+}
+
+function injectPixels(ad, pixels, scripts) {
+  if (!pixels && !scripts) {
+    return ad;
+  }
+
+  let trackedAd = ad;
+  if (pixels) {
+    pixels.forEach(pixel => {
+      const tracker = utils.createTrackPixelHtml(pixel);
+      trackedAd += tracker;
+    });
+  }
+
+  if (scripts) {
+    scripts.forEach(script => {
+      const tracker = `<script src="${script}"></script>`;
+      trackedAd += tracker;
+    });
+  }
+
+  return trackedAd;
+}
+
+function getScreenParams() {
+  return `${window.screen.width}x${window.screen.height}@${window.devicePixelRatio}`;
+}
+
+function getBids(bids) {
+  const bidArr = bids.map(bid => {
+    const bidId = bid.bidId;
+
+    let mediaType = '';
+    const mediaTypes = Object.keys(bid.mediaTypes)
+    switch (mediaTypes[0]) {
+      case 'video':
+        mediaType = 'v';
+        break;
+
+      case 'native':
+        mediaType = 'n';
+        break;
+
+      case 'audio':
+        mediaType = 'a';
+        break;
+
+      default:
+        mediaType = 'b';
+        break;
+    }
+
+    let adUnitCode = `,c=${bid.adUnitCode}`;
+    if (bid.params.code) {
+      adUnitCode = `,c=${encodeURIComponent(bid.params.code)}`;
+    }
+    if (bid.params.adunitId) {
+      adUnitCode = `,u=${encodeURIComponent(bid.params.adunitId)}`;
+    }
+
+    return `${bidId}:t=${mediaType},s=${serializeSizes(bid.sizes)}${adUnitCode}`;
+  });
+
+  return bidArr.join(';');
+};
+
+function getEndpointsGroups(bidRequests) {
+  let endpoints = [];
+  const getEndpoint = bid => {
+    if (bid.params.test) {
+      return `https://mock-bapi.userreport.com/v2/${bid.params.publisherId}/bid`;
+    }
+
+    if (bid.params.endpoint) {
+      return `${bid.params.endpoint}${bid.params.publisherId}/bid`;
+    }
+
+    return `https://bapi.userreport.com/v2/${bid.params.publisherId}/bid`;
+  }
+  bidRequests.forEach(bid => {
+    const exist = endpoints.filter(item => item.endpoint.indexOf(bid.params.endpoint) > -1)[0];
+    if (exist) {
+      exist.bids.push(bid);
+    } else {
+      endpoints.push({
+        endpoint: getEndpoint(bid),
+        bids: [bid]
+      });
+    }
+  });
+
+  return endpoints;
+}
+
+function isBidRequestValid(bid) {
+  const isPublisherIdExist = !!bid.params.publisherId;
+  const isOneMediaType = Object.keys(bid.mediaTypes).length === 1;
+
+  return isPublisherIdExist && isOneMediaType;
+}
+
+function buildRequests(bidRequests, bidderRequest) {
+  const data = {
+    med: encodeURIComponent(window.location.href),
+    auid: bidderRequest.auctionId,
+    ref: document.referrer,
+    dnt: utils.getDNT() ? 1 : 0,
+    sr: getScreenParams()
+  };
+
+  const consentData = getRawConsentString(bidderRequest.gdprConsent);
+  data.iab_consent = consentData;
+
+  let options = {
+    withCredentials: true
+  };
+
+  const isConsent = getIabConsentString(bidderRequest);
+  if (isConsent) {
```

Comment:
can you add a bidder config option here and document it so that pubs who aren't comfortable with this functionality can disable the dsu part of your adapter? I think it is okay to default to on but should be on the bidder params section of the documentation at http://prebid.org/dev-docs/bidders.html

Reply:
noting from the module rules:

"Bidder modules must not(*) make requests to endpoints for functionality other than auctions without:
Disclosure (See the disclosure section below.)
Ability for the publisher to control the additional functionality."

Reply:
https://docs.google.com/document/d/1kK0ifXV-vzyV7VchPKoyGuz375H7fzbTJFt_oWeCiRo/

Reply:
Hi @patmmccann. We are not making any requests to endpoints related to DSU. As I can see in the docs, this rule is applied only for requests to endpoints for functionality other than auctions. In our case DSU just identify user and is generated on client side, this is a part of our workflow. Also it uses local storage only if consent is given.

Reply:
I think publishers being able to control this additional functionality (generate and pass dsu) aligns quite well with module rules. The idea is some publishers are subject to  regulatory scrutiny or have stronger internal standards and may wish to control this functionality at a more detailed level than (isconsent) and may wish to check for consent for various tcf2 purposes.

Reply:
Hi @patmmccann. Where is better place to set this parameter? I can add this option to params section to every bid, but does it makes sense as far this DSU flag is browser specific but not bid specific? Does it means what I should take into account parameter only from first bit or what do you could suggest?

Reply:
I imagine a publisher would do something like: `var adUnits = [{
      code: 'test-div',
      sizes: [[300, 600]],
      bids: [{
          bidder: 'apstream',
          params: { 
              dsu: 'off' //defaults to on
          }
      }]
    }];` when they want to override it

Reply:
There's precedence for bidder-specific global params. e.g.
```
pbjs.setConfig({
   apstream: {dsu: false}
});
```
Your code can do a `getConfig` to look for this optional param.